tools/downloader/topologies.py

```diff
@@ -0,0 +1,196 @@
+
+        if useIR:
+            return m
+
+        if '--mean_values' in self.mo_args:
+            mean = self.mo_args['--mean_values']
+            mean = mean[mean.find('[') + 1:mean.find(']')].split(',')
+            mean = [float(val) for val in mean]
+            m.setInputMean(mean)
+
+        if '--scale_values' in self.mo_args:
+            scale = self.mo_args['--scale_values']
+            scale = 1.0 / float(scale[scale.find('[') + 1:scale.find(']')])
+            m.setInputScale(scale)
+
+        if '--input_shape' in self.mo_args:
+            shape = self.mo_args['--input_shape']
+            shape = shape[shape.find('[') + 1:shape.find(']')].split(',')
+            shape = [int(val) for val in shape]
+            if len(shape) == 4:
+                if self.framework == 'tf':  # NHWC
+                    w, h = shape[2], shape[1]
+                else:  # NCHW
+                    w, h = shape[3], shape[2]
+            m.setInputSize(w, h)
+
+        if '--reverse_input_channels' in self.mo_args:
+            m.setInputSwapRB(True)
+
+        return m
+
+
+    def getIENetwork(self):
+        '''
+        Creates openvino.inference_engine.IENetwork instance.
+        Model Optimizer is lauched to create OpenVINO Intermediate Representation (IR).
+        '''
+        from openvino.inference_engine import IENetwork
+        xmlPath, binPath = self.getIR()
+        return IENetwork(xmlPath, binPath)
+
+
+for _topology in common.load_models(argparse.Namespace(config=None)):
+    _name = _topology.name.replace('-', '_').replace('.', '_')
+    _download_dir = _cache / _topology.subdirectory
+    _mo_args = {}
+    if _topology.mo_args:
+        for _arg in _topology.mo_args:
+            _tokens = _arg.split('=')
+            _mo_args[_tokens[0]] = _tokens[1] if len(_tokens) == 2 else None
+
+    # Resolve weights and text files names.
+    if _topology.framework == 'dldt':
+        assert(len(_topology.files) >= 2), 'Expected to have at least two files in IR format'
+        _paths = []
+        # DLDT topologies usually come with multiple precision configuration files.
+        for i in range(2):
+            # Get only basename and add precision marker
+            _path = str(_topology.files[i].name)
+            _pos = _path.find('/')
+            if _path[:_pos] in common.KNOWN_PRECISIONS:
+                _path = _precision_marker + _path[_pos:]
+            _paths.append(_path)
+
+        _config_path, _model_path = _paths if _paths[0].endswith('.xml') else _paths[::-1]
+    else:
+        # Get paths from Model Optimizer arguments
+        if '--input_model' in _mo_args:
+            _model_path = _mo_args['--input_model'].replace('$dl_dir/', '')
+            _config_path = ''  # If there is a text file - we will set it next
+
+        for key in ['--input_proto', '--input_symbol']:
+            if key in _mo_args:
+                _config_path = _mo_args[key].replace('$dl_dir/', '')
```

Comment:
This is very hacky, and I don't like it.

* There might be variables other than `dl_dir`.
* There might be different variable syntax (`string.Template`, which we use, supports `${var}` as well as `$var`).
* This logic will fail for models that have to be converted to ONNX first.
* It will also fail for TensorFlow checkpoint-based models.

In general, the whole practice of parsing Model Optimizer arguments is flawed, because you have no guarantee that any arguments you know about now are sufficient to fully describe any future model.

I don't think the model configs currently contain enough information to enable safe automated usage of the models in source formats. I think it might be useful to provide this information, but that is a feature in its own right. For now, though, I think you should remove any such usage.

Reply:
Replaced to string's substitution as it's implemented in https://github.com/opencv/open_model_zoo/blob/develop/tools/downloader/converter.py

Reply:
But this doesn't address my point at all... You just made your ad-hoc logic a little bit more ad-hoc.